### PR TITLE
Add binary focal cross entropy loss function

### PR DESCRIPTION
**Pull request by AtomicVar**
## Proposed changes

This PR adds the **Binary Focal Cross-entropy Loss**, which is introduced by [Lin et al., 2018](https://arxiv.org/pdf/1708.02002.pdf) and often used in object detection tasks.

The interface is nearly the same with TensorFlow's [tf.keras.losses.BinaryFocalCrossentropy](https://www.tensorflow.org/api_docs/python/tf/keras/losses/BinaryFocalCrossentropy) and the test ground-truth values are calculated by it to ensure correctness.

This implementation supports both logits and probabilities as input (like the normal `binary_cross_entropy`). It also supports **class balancing** in [Lin et al., 2018](https://arxiv.org/pdf/1708.02002.pdf), like TF's `BinaryFocalCrossentropy`.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by awni** [comment]
Thanks for the contribution! We decided against adding this to MLX core in #336 and #324 as it's a bit niche. We may revisit in the future.

For future reference, it's a good idea to open an issue about possible new features before or in tandem to starting a PR so we can let you know if there is ongoing work / plans already / not for core.
